builder_test.go

```diff
@@ -277,3 +281,97 @@ func TestPath(t *testing.T) {
 		})
 	}
 }
+
+func TestPreRequestHook(t *testing.T) {
+	addBodyDigestHeader := func(r *http.Request) error {
+		if r.Body == nil {
+			return nil
+		}
+
+		body, err := r.GetBody()
+		if err != nil {
+			return err
+		}
+
+		bodyBytes, err := io.ReadAll(body)
+		if err != nil {
+			return err
+		}
+
+		r.Header.Add("digest", fmt.Sprintf("%x", sha256.Sum256(bodyBytes)))
+		return nil
+	}
+
+	t.Run("body-hash-header", func(t *testing.T) {
+		body := []byte("hello world")
+		b := requests.URL("https://example").
+			BodyBytes(body).
+			PreRequestHook(addBodyDigestHeader)
```

Comment:
I would take a PR for a digest header RoundTripper if you can show that it's a common need. I don't need one ATM, but I needed one for Python years ago and I suspect it comes up a lot.